tornado/routing.py

```diff
@@ -0,0 +1,335 @@
+                return route
+
+        elif callable(target):
+            return _CallableAdapter(
+                partial(target, **target_params), request.connection
+            )
+
+        return None
+
+
+class ReversibleRuleRouter(ReversibleRouter, RuleRouter):
+    def __init__(self, rules=None):
+        self.named_rules = {}
+        super(ReversibleRuleRouter, self).__init__(rules)
+
+    def process_rule(self, rule):
+        rule = super(ReversibleRuleRouter, self).process_rule(rule)
+
+        if rule.name:
+            if rule.name in self.named_rules:
+                app_log.warning(
+                    "Multiple handlers named %s; replacing previous value",
+                    rule.name)
+            self.named_rules[rule.name] = rule
+
+        return rule
+
+    def reverse_url(self, name, *args):
+        if name in self.named_rules:
+            return self.named_rules[name].matcher.reverse(*args)
+
+        for rule in self.rules:
+            if isinstance(rule.target, ReversibleRouter):
+                reversed_url = rule.target.reverse_url(name, *args)
+                if reversed_url is not None:
+                    return reversed_url
+
+        return None
+
+
+class Rule(object):
+    def __init__(self, matcher, target, target_kwargs=None, name=None):
+        self.matcher = matcher
+        self.target = target
+        self.target_kwargs = target_kwargs if target_kwargs else {}
+        self.name = name
+
+    def reverse(self, *args):
+        if isinstance(self.matcher, Matcher):
+            return self.matcher.reverse(*args)
+        return None
+
+    def __repr__(self):
+        return '%s(%r, %s, kwargs=%r, name=%r)' % \
+               (self.__class__.__name__, self.matcher,
+                self.target, self.target_kwargs, self.name)
+
+
+class Matcher(object):
+    """A complex matcher can be represented as an instance of some
+    `Matcher` subclass. It must implement ``__init__`` (for example
+    to initialize a regex pattern) and ``__call__`` (which is called
+    with a `httpserver.HTTPRequest` argument).
+    """
+
+    def __call__(self, request):
+        """Matches an instance against the request.
+
+        :returns a dict of parameters (handler_kwargs, path_args, path_kwargs)
+        to be passed to the target handler"""
+        raise NotImplementedError()
+
+    def reverse(self, *args):
+        """Reconstruct URL from matcher instance"""
+        return None
+
+
+ANY_MATCHES = lambda request: {}
+
+
+class host_matches(Matcher):
+    def __init__(self, host_pattern):
+        if isinstance(host_pattern, basestring_type):
+            if not host_pattern.endswith("$"):
+                host_pattern += "$"
+            self.host_pattern = re.compile(host_pattern)
+        else:
+            self.host_pattern = host_pattern
+
+    def __call__(self, request):
+        if not hasattr(request, '_host'):
+            request._host = httputil.split_host_and_port(request.host.lower())[0]
```

Comment:
A typical usage of Rules/Matchers _in application context_ will be roughly [like this](https://github.com/tornadoweb/tornado/pull/1806/files#diff-14a4caccf0d2dcc65cc129ae9186ef3eR100).
Target may be a nested Router, but this could be neccessary only for some very advanced cases.

A simpler solution is to specify a list of targets (a list of rules) as in example from the tests. They are implicitly wrapped in an ApplicationRouter, which preserves the application context.

If we move to SpecificRouter([non-typed rules]) scheme, then we must implicitly pass an Application instance (if it is present) to these routers somehow. I guess this could be done, but it is a bit more complicated.